expression/new_constant_propagation.go

```diff
@@ -0,0 +1,162 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package expression
+
+import (
+	"bytes"
+
+	"github.com/pingcap/tidb/mysql"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/sessionctx/stmtctx"
+	"github.com/pingcap/tidb/types"
+)
+
+// exprSet is a Set container for expressions, each expression in it is unique.
+// `tombstone` is deleted mark, if tombstone[i] is false, data[i] is invalid.
+// `index` use expr.HashCode() as key, to implement the unique property.
+type exprSet struct {
+	data       []Expression
+	tombstone  []bool
+	index      map[string]struct{}
+	constfalse bool
+}
+
+func (s *exprSet) Append(e Expression) bool {
+	if _, ok := s.index[string(e.HashCode(nil))]; ok {
+		return false
+	}
+
+	s.data = append(s.data, e)
+	s.tombstone = append(s.tombstone, false)
+	s.index[string(e.HashCode(nil))] = struct{}{}
+	return true
+}
+
+// Slice returns the valid expressions in the exprSet, this function has side effect.
+func (s *exprSet) Slice() []Expression {
+	if s.constfalse {
+		return []Expression{&Constant{
+			Value:   types.NewDatum(false),
+			RetType: types.NewFieldType(mysql.TypeTiny),
+		}}
+	}
+
+	idx := 0
+	for i := 0; i < len(s.data); i++ {
+		if !s.tombstone[i] {
+			s.data[idx] = s.data[i]
+			idx++
+		}
+	}
+	return s.data[:idx]
+}
+
+func (s *exprSet) SetConstFalse() {
+	s.constfalse = true
+}
+
+func newExprSet(conditions []Expression) *exprSet {
+	var exprs exprSet
+	exprs.data = make([]Expression, 0, len(conditions))
+	exprs.tombstone = make([]bool, 0, len(conditions))
+	exprs.index = make(map[string]struct{}, len(conditions))
+	for _, v := range conditions {
+		exprs.Append(v)
+	}
+	return &exprs
+}
+
+type pgSolver2 struct{}
+
+// PropagateConstant propagate constant values of deterministic predicates in a condition.
+func (pgSolver2) PropagateConstant(ctx sessionctx.Context, conditions []Expression) []Expression {
+	exprs := newExprSet(conditions)
+	fixPoint(ctx, exprs)
+	return exprs.Slice()
+}
+
+// fixPoint is the core of the constant propagation algorithm.
+// It will iterate the expression set over and over again, pick two expressions,
+// apply one to another.
+// If new conditions can be infered, they will be append into the expression set.
+// Until no more conditions can be infered from the set, the algorithm finish.
+func fixPoint(ctx sessionctx.Context, exprs *exprSet) {
+	for {
+		saveLen := len(exprs.data)
+		iterOnce(ctx, exprs)
+		if saveLen == len(exprs.data) {
+			break
+		}
+	}
+	return
+}
+
+// iterOnce picks two expressions from the set, try to propagate new conditions from them.
+func iterOnce(ctx sessionctx.Context, exprs *exprSet) {
+	for i := 0; i < len(exprs.data); i++ {
+		if exprs.tombstone[i] {
+			continue
+		}
+		for j := 0; j < len(exprs.data); j++ {
```

Comment:
would start j from i+1 and double the logic of each rule outperform this full loop? though it looks uglier...

Reply:
The solve function does not handle both direction:
solve(i, j)  = propagate use data[i] as rule, and apply it on data[j]
Even solve(5, 6) is handled, solve(6, 5) is necessary, so j need to start from 0

Reply:
Yes, I mean double the logic of solve function to handle both direction also. Just though it should involve less function call then. Anyway, start j from 0 is also OK for me.

Reply:
The performance of this algorithm may be not very good, I guess, I want to use it for my special partition pruning solution first, so it's not optimized for general purpose.
Let solve function support both direction would lead to more code and make writing rules harder, so I think it doesn't worth the work.